File: boj/boj_10830.java
package com.ssafy.boj;

import java.util.Scanner;

public class Main {

	public static int N;
	public static long B;
	public static int A[][];
	public static int ans[][];

	public static void main(String[] args) {
		processInput();
		ans = power(A, B);
		printAnswer();
	}

	public static void processInput() {
		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		B = sc.nextLong();

		A = new int[N][N];
		ans = new int[N][N];

		for (int i = 0; i < N; ++i) {
			for (int j = 0; j < N; ++j) {
				A[i][j] = sc.nextInt();
			}
			ans[i][i] = 1;
		}
	}

	public static void printAnswer() {
		for (int i = 0; i < N; ++i) {
			for (int j = 0; j < N; ++j) {
				System.out.print(ans[i][j] + " ");
			}
			System.out.println();
		}
	}

	public static int[][] power(int[][] A, long B) {
		int[][] ret = createIdentity();

		if (B == 1) {
			ret = product(ret, A);
		} else {
			ret = power(A, B / 2);
			ret = product(ret, ret);

			if (B % 2 == 1)
				ret = product(ret, A);
		}

		return ret;
	}

	public static int[][] product(int[][] m1, int[][] m2) {
		int[][] ret = new int[N][N];

		for (int i = 0; i < N; ++i) {
			for (int j = 0; j < N; ++j) {
				for (int k = 0; k < N; ++k) {
					ret[i][j] += (m1[i][k] * m2[k][j]);
				}
				ret[i][j] %= 1000;
			}
		}

		return ret;
	}

	public static int[][] createIdentity() {
		int[][] ret = new int[N][N];
		for (int i = 0; i < N; ++i)
			ret[i][i] = 1;
		return ret;
	}
}

File: boj/boj_10830.cpp
#include <iostream>

void applyPower(int ans[5][5], int A[5][5], long long n);
void product(int A[5][5], int B[5][5]);

void processInput();
void printAnswer();

long long N, B;
int ans[5][5];
int A[5][5];

int main()
{
	processInput();
	applyPower(ans, A, B);
	printAnswer();

	return 0;
}

void applyPower(int ans[5][5], int A[5][5], long long n)
{
	if (n == 1)
	{
		product(ans, A);
	}
	else
	{
		applyPower(ans, A, n / 2);
		product(ans, ans);

		if (n % 2 == 1) product(ans, A);
	}
}

void product(int A[5][5], int B[5][5])
{
	int temp[5][5];
	for (int i = 0; i < N; ++i)
	{
		for (int j = 0; j < N; ++j)
		{
			temp[i][j] = 0;
			for (int k = 0; k < N; ++k)
			{
				temp[i][j] += (A[i][k] * B[k][j]);
			}
			temp[i][j] %= 1000;
		}
	}

	for (int i = 0; i < N; ++i)
	{
		for (int j = 0; j < N; ++j)
		{
			A[i][j] = temp[i][j];
		}
	}
}

void processInput()
{
	std::cin >> N >> B;
	for (int i = 0; i < N; ++i)
	{
		for (int j = 0; j < N; ++j)
		{
			std::cin >> A[i][j];
		}
		ans[i][i] = 1;
	}
}

void printAnswer()
{
	for (int i = 0; i < N; ++i)
	{
		for (int j = 0; j < N; ++j)
		{
			std::cout << ans[i][j] << " ";
		}
		std::cout << "\n";
	}
}